tests/optimize/test_rotosolve.py

```diff
@@ -0,0 +1,293 @@
+
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.utils import _flatten, unflatten
+
+from pennylane.optimize import RotosolveOptimizer
+
+
+def expand_num_freq(num_freq, param):
+    if np.isscalar(num_freq):
+        num_freq = [num_freq] * len(param)
+    expanded = []
+    for _num_freq, par in zip(num_freq, param):
+        if np.isscalar(_num_freq) and np.isscalar(par):
+            expanded.append(_num_freq)
+        elif np.isscalar(_num_freq):
+            expanded.append(np.ones_like(par) * _num_freq)
+        elif np.isscalar(par):
+            raise ValueError(f"{num_freq}\n{param}\n{_num_freq}\n{par}")
+        elif len(_num_freq) == len(par):
+            expanded.append(_num_freq)
+        else:
+            raise ValueError()
+    return expanded
+
+
+def successive_params(par1, par2):
+    """Return a list of parameter configurations, successively walking from
+    par1 to par2 coordinate-wise."""
+    par1_flat = np.fromiter(_flatten(par1), dtype=float)
+    par2_flat = np.fromiter(_flatten(par2), dtype=float)
+    walking_param = []
+    for i in range(len(par1_flat) + 1):
+        walking_param.append(unflatten(np.append(par2_flat[:i], par1_flat[i:]), par1))
+    return walking_param
+
+
+@pytest.mark.parametrize(
+    "fun, param, num_freq",
+    zip(
+        [
+            lambda x: np.sin(x),
+            lambda x: np.sin(x),
+            lambda x, y: np.sin(x) * np.sin(y),
+            lambda x, y: np.sin(x) * np.sin(y[0]) * np.sin(y[1]),
+        ],
+        [[0.5], [0.5], [0.5, 0.2], [0.5, [0.2, 0.4]]],
+        [[], [1, 1], [1], [1, 1, [1, 2]]],
+    ),
+)
+def test_wrong_len_num_frequencies(fun, param, num_freq):
+    """Test that an error is raised for a different number of
+    numbers of frequencies than number of function arguments."""
+
+    opt = qml.RotosolveOptimizer()
+
+    with pytest.raises(ValueError, match="The number of the provided numbers of frequencies"):
+        opt.step(fun, *param, num_frequencies=num_freq)
+
+
+@pytest.mark.parametrize(
+    "fun, param, num_freq",
+    zip(
+        [
+            lambda x: np.sin(x),
+            lambda x: np.sin(x),
+            lambda x, y: np.sin(x) * np.sin(y),
+            lambda x, y: np.sin(x) * np.sin(y[0]) * np.sin(y[1]),
+        ],
+        [[0.5], [0.5], [0.5, 0.2], [0.5, [0.2, 0.4]]],
+        [[[1, 1]], [[]], [[1], [1, 1]], [[1], [1]]],
+    ),
+)
+def test_wrong_num_of_num_frequencies_per_parameter(fun, param, num_freq):
+    """Test that an error is raised for a different number of
+    numbers of frequencies than number of function arguments."""
+
+    opt = qml.RotosolveOptimizer()
+
+    with pytest.raises(ValueError, match="The number of the numbers of frequencies"):
+        opt.step(fun, *param, num_frequencies=num_freq)
+
+
+classical_functions = [
+    lambda x: np.sin(x + 0.124) * 2.5123,
+    lambda x: -np.cos(x[0] + 0.12) * 0.872 + np.sin(x[1] - 2.01) - np.cos(x[2] - 1.35) * 0.111,
```

Comment:
Good idea to test an array parameter here. I have a question; should there be a QNode integration test?
 
In particular, I'm thinking of a QNode integration test that uses a template with a multi-dim array input (like StronglyEntanglingLayers) just to make sure it works.

Reply:
I was thinking about this and yes, maybe it's good for completeness. I could not come up with a case where this would go wrong while the existing tests with classical functions go through, but maybe it's also nice in order to have a compatibility test between qnode signature and Rotosolve :)